pipenv/environment.py

```diff
@@ -351,33 +351,84 @@ def get_outdated_packages(self, pre=False):
+        if pkg:
+            packages = [p for p in packages if p.key == pkg]
+        dist_index = build_dist_index(packages)
         tree = sorted_tree(construct_tree(dist_index))
         branch_keys = set(r.key for r in flatten(tree.values()))
-        nodes = [p for p in tree.keys() if p.key not in branch_keys]
+        if pkg is not None:
+            nodes = [p for p in tree.keys() if p.key == pkg]
+        else:
+            nodes = [p for p in tree.keys() if p.key not in branch_keys]
         key_tree = dict((k.key, v) for k, v in tree.items())
-        get_children = lambda n: key_tree.get(n.key, [])
-
-        def aux(node, parent=None, chain=None):
-            if chain is None:
-                chain = [node.project_name]
 
-            d = node.as_dict()
-            if parent:
-                d['required_version'] = node.version_spec if node.version_spec else 'Any'
-            else:
-                d['required_version'] = d['installed_version']
-
-            d['dependencies'] = [
-                aux(c, parent=node, chain=chain+[c.project_name])
-                for c in get_children(node)
-                if c.project_name not in chain
-            ]
+        return [self._get_requirements_for_package(p, key_tree) for p in nodes]
 
-            return d
-        return [aux(p) for p in nodes]
+    @classmethod
+    def reverse_dependency(cls, node):
+        new_node = {
+            "package_name": node["package_name"],
+            "installed_version": node["installed_version"],
+            "required_version": node["required_version"]
+        }
+        for dependency in node.get("dependencies", []):
+            for dep in cls.reverse_dependency(dependency):
+                new_dep = dep.copy()
+                new_dep["parent"] = (node["package_name"], node["installed_version"])
+                yield new_dep
+        yield new_node
+
+    def reverse_dependencies(self):
+        from vistir.misc import unnest
+        rdeps = {}
+        for req in self.get_package_requirements():
+            for d in self.reverse_dependency(req):
+                name = d["package_name"]
+                pkg = {
+                    name: {
+                        "installed": d["installed_version"],
+                        "required": d["required_version"]
+                    }
+                }
+                if d.get("parent"):
+                    pkg[name]["parents"] = list(d["parent"])
+                if rdeps.get(name):
+                    if rdeps[name].get("parents"):
+                        rdeps[name]["parents"].append(pkg[name]["parents"])
+                    else:
+                        rdeps[name].update(pkg[name])
+                else:
+                    rdeps.update(pkg)
+        for k in list(rdeps.keys()):
+            entry = rdeps[k]
+            if entry.get("parents"):
+                rdeps[k]["parents"] = set([p for p in unnest(entry["parents"])])
```

Comment:
right, because it's not meant to get the parent dependencies, it's meant to get all of the reverse dependencies.  That is, every single dependency that is in the chain of dependencies demanding this specific package.

That is, if you installed `A`, and `A -> B -> C`, the point of `reverse_dependencies` is to tell you about both `B` and `C`, this way you can figure out everything in the way of removing or upgrading a specific package.

The implementation reason for this is that if you want to introduce or update a new package, and that package in turn needs to make a modification to another package already locked in the lockfile, you can't safely do that unless you know about all of the things that are in the dependency chain which rely on the version which has been pinned.  That means you need a way to get that information.